### PR TITLE
Improve documentation of ``HamiltonianGate`` and ``UnitaryGate``

### PR DESCRIPTION
**Pull request by adrinjalali**
I first noticed a hyperlink is not resolved on this page: https://qiskit.org/documentation/stubs/qiskit.extensions.UnitaryGate.html?highlight=unitarygate#qiskit.extensions.UnitaryGate

while trying to fix that, also realized some docs were missing from the methods attached to `QuantumCircuit` and that `HamiltonianGate` doesn't refer to the method the same way `UnitaryGate` does.

The docs added to the functions are copy/pasted from the gate docs, I haven't written anything new here.

### CODE REVIEW

File: qiskit/extensions/hamiltonian_gate.py [hunk 1]
```diff
@@ -27,9 +27,12 @@
...
 class HamiltonianGate(Gate):
-    """Class for representing evolution by a Hermitian Hamiltonian operator as a gate. This gate
-    resolves to a UnitaryGate U(t) = exp(-1j * t * H), which can be decomposed into basis gates if
-    it is 2 qubits or less, or simulated directly in Aer for more qubits."""
+    """Class for representing evolution by a Hermitian Hamiltonian operator as
+    a gate. This gate resolves to a :class:`UnitaryGate` as :math:`U(t) =
+    exp(-1j * t * H)`, which can be decomposed into basis gates if it is 2
+    qubits or less, or simulated directly in Aer for more qubits. Note that you
+    can also directly use :meth:`~qiskit.circuit.QuantumCircuit.hamiltonian`.
+    """
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:35]: With a single header line this renders nicely in the docs 🙂 Also in math mode we don't need to use Python notation 😄 
```suggestion
    """Class for representing evolution by a Hamiltonian operator as a gate.
    
    This gate resolves to a :class:`.UnitaryGate` as :math:`U(t) =
    exp(-i t H)`, which can be decomposed into basis gates if it is 2
    qubits or less, or simulated directly in Aer for more qubits. Note that you
    can also directly use :meth:`.QuantumCircuit.hamiltonian`.
    """
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 2]
```diff
@@ -127,7 +130,25 @@ def validate_parameter(self, parameter):
 
 
 def hamiltonian(self, operator, time, qubits, label=None):
-    """Apply hamiltonian evolution to qubits."""
+    """Apply hamiltonian evolution to qubits.
+
+    This gate resolves to a :class:`~qiskit.extensions.UnitaryGate` as
+    :math:`U(t) = exp(-1j * t * H)`, which can be decomposed into basis gates
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:136]:
```suggestion
    This gate resolves to a :class:`.UnitaryGate` as
    :math:`U(t) = exp(-i t H)`, which can be decomposed into basis gates
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 3]
```diff
@@ -127,7 +130,25 @@ def validate_parameter(self, parameter):
...
+    if it is 2 qubits or less, or simulated directly in Aer for more qubits.
+
+    Args:
+        operator (matrix or Operator): a hermitian operator.
+        time (float): time evolution parameter.
+        qubits (Union[int, Tuple[int]]): The circuit qubits to apply the
+            transformation to.
+        label (str): unitary name for backend [Default: None].
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:144]: Could you put the typehints in the function definition instead of the docs here? 🙂

Comment (adrinjalali) [reply qiskit/extensions/hamiltonian_gate.py:144]: Almost none of the other methods here or in `unitary.py` have type hints though. Moving these to type hints would make it quite inconsistent, and I think it'd be better as a candidate for a separate PR, don't you think?

File: qiskit/extensions/hamiltonian_gate.py [hunk 4]
```diff
@@ -127,7 +131,25 @@ def validate_parameter(self, parameter):
...
-    """Apply hamiltonian evolution to qubits."""
+    """Apply hamiltonian evolution to qubits.
+
+    This gate resolves to a :class:`.UnitaryGate` as :math:`U(t) = exp(-i t H)`,
+    which can be decomposed into basis gates if it is 2 qubits or less, or
+    simulated directly in Aer for more qubits.
+
+    Args:
+        operator (matrix or Operator): a hermitian operator.
+        time (float): time evolution parameter.
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:142]: This code also allows `ParameterExpression` as type (see eg. L91 and following which cover this case), could you add that type in the initializer and this method here? 🙂

File: qiskit/extensions/hamiltonian_gate.py (hunk 3 above)
Comment (Cryoris) [reply qiskit/extensions/hamiltonian_gate.py:145]: Ok makes sense 👍🏻

**Review by Cryoris — COMMENTED** [review]
One minor comment, otherwise LGTM 🙂

File: qiskit/extensions/hamiltonian_gate.py (hunk 4 above)
Comment (adrinjalali) [reply qiskit/extensions/hamiltonian_gate.py:142]: does that apply to `operator` or `time`? I can't easily figure out where `validate_parameter` is used and the docstring is not very clear.

Comment (Cryoris) [reply qiskit/extensions/hamiltonian_gate.py:142]: That applies to `time` 🙂

File: qiskit/extensions/hamiltonian_gate.py [hunk 5]
```diff
@@ -127,7 +132,25 @@ def validate_parameter(self, parameter):
...
-    """Apply hamiltonian evolution to qubits."""
+    """Apply hamiltonian evolution to qubits.
+
+    This gate resolves to a :class:`.UnitaryGate` as :math:`U(t) = exp(-i t H)`,
+    which can be decomposed into basis gates if it is 2 qubits or less, or
+    simulated directly in Aer for more qubits.
+
+    Args:
+        operator (matrix or Operator): a hermitian operator.
+        time (float, ndarray or ParameterExpression): time evolution parameter.
```
Comment (adrinjalali) [inline qiskit/extensions/hamiltonian_gate.py:143]: does this look better @Cryoris ?

File: qiskit/extensions/hamiltonian_gate.py [hunk 6]
```diff
@@ -27,16 +27,21 @@
...
+    """
 
     def __init__(self, data, time, label=None):
         """Create a gate from a hamiltonian operator and evolution time parameter t
 
         Args:
             data (matrix or Operator): a hermitian operator.
-            time (float): time evolution parameter.
+            time (float, ndarray or ParameterExpression): time evolution
+                parameter.
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:44]:
```suggestion
            time (float or ParameterExpression): time evolution parameter.
```

File: qiskit/extensions/hamiltonian_gate.py (hunk 5 above)
Comment (Cryoris) [reply qiskit/extensions/hamiltonian_gate.py:143]:
```suggestion
        time (float or ParameterExpression): time evolution parameter.
```

**Review by Cryoris — APPROVED** [review]
Thanks for the updates!

**Comment by adrinjalali** [comment]
@Cryoris I had `ndarray` since it's accepted by `validate_parameter`, should it then not be accepted there?

**Pull request merged**